rust-runtime/aws-smithy-types/src/timeout.rs

```diff
@@ -0,0 +1,356 @@
+///
+/// # Example
+///
+/// ```rust
+/// # use std::time::Duration;
+///
+/// # fn main() {
+/// use aws_smithy_types::timeout::TimeoutConfig;
+/// let timeout_config = TimeoutConfig::new()
+///     .with_api_call_timeout(Duration::from_secs(2))
+///     .with_api_call_attempt_timeout(Duration::from_secs_f32(0.5));
+///
+/// assert_eq!(
+///     timeout_config.api_call_timeout(),
+///     Some(Duration::from_secs(2))
+/// );
+///
+/// assert_eq!(
+///     timeout_config.api_call_attempt_timeout(),
+///     Some(Duration::from_secs_f32(0.5))
+/// );
+/// # }
+/// ```
+#[derive(Debug, Clone, PartialEq, Default)]
+pub struct TimeoutConfig {
+    connect_timeout: Option<Duration>,
+    tls_negotiation_timeout: Option<Duration>,
+    read_timeout: Option<Duration>,
+    api_call_attempt_timeout: Option<Duration>,
+    api_call_timeout: Option<Duration>,
+}
+
+impl TimeoutConfig {
+    /// Create a new `TimeoutConfig` with no timeouts set
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Create a new [TimeoutConfigBuilder]
+    /// A limit on the amount of time after making an initial connect attempt on a socket to complete the connect-handshake.
+    pub fn connect_timeout(&self) -> Option<Duration> {
+        self.connect_timeout
+    }
+
+    /// A limit on the amount of time a TLS handshake takes from when the `CLIENT HELLO` message is
+    /// sent to the time the client and server have fully negotiated ciphers and exchanged keys.
+    pub fn tls_negotiation_timeout(&self) -> Option<Duration> {
+        self.tls_negotiation_timeout
+    }
+
+    /// A limit on the amount of time an application takes to attempt to read the first byte over an
+    /// established, open connection after write request. This is also known as the
+    /// "time to first byte" timeout.
+    pub fn read_timeout(&self) -> Option<Duration> {
+        self.read_timeout
+    }
+
+    /// A limit on the amount of time it takes for the first byte to be sent over an established,
+    /// open connection and when the last byte is received from the service for a single attempt.
+    /// If you want to set a timeout for an entire request including retry attempts,
+    /// use [`TimeoutConfig::api_call_timeout`] instead.
+    pub fn api_call_attempt_timeout(&self) -> Option<Duration> {
+        self.api_call_attempt_timeout
+    }
+
+    /// A limit on the amount of time it takes for request to complete. A single request may be
+    /// comprised of several attemps depending on an app's [`RetryConfig`](super::retry::RetryConfig). If you want
+    /// to control timeouts for a single attempt, use [`TimeoutConfig::api_call_attempt_timeout`].
+    pub fn api_call_timeout(&self) -> Option<Duration> {
+        self.api_call_timeout
+    }
+
+    /// Generate a human-readable list of the timeouts that are currently set. This is used for
+    /// logging and error reporting.
+    pub fn list_of_set_timeouts(&self) -> Vec<&'static str> {
```

Comment:
I'm not sure this gets us much beyond the existing debug impl (which works without any allocations)